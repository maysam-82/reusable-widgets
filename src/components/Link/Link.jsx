import React from 'react';

const Link = ({ href, className, children }) => {
	const onClickHandler = (event) => {
		if (event.metaKey || event.ctrlKey) {
			return;
		}

		event.preventDefault();

		// Change window.location.pathname to `href` value
		window.history.pushState({}, '', href);

		// Creating custom events for `popstate`
		// `PopStateEvent` is an event handler for the `popstate` event on the window.
		// https://developer.mozilla.org/en-US/docs/Web/API/PopStateEvent
		const navEvent = new PopStateEvent('popstate');
		window.dispatchEvent(navEvent);
	};
	return (
		<a href={href} className={className} onClick={onClickHandler}>
			{children}
		</a>
	);
};

export default Link;
