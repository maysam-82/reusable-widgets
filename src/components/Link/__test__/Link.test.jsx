import React from 'react';
import { shallow } from 'enzyme';
import Link from '../Link';

describe('Link component', () => {
	const testRoutePath = '/test-route';
	const component = shallow(
		<Link href={testRoutePath} className="test-class">
			Test Component
		</Link>
	);
	it('should render Link component', () => {
		expect(component).toMatchSnapshot();
	});
	describe('When user clicks on `Link` component and does not hold `Crtl` key', () => {
		const event = {
			metaKey: false,
			ctrlKey: false,
			preventDefault: jest.fn(),
		};
		beforeEach(() => {
			component.find('a').simulate('click', event);
		});
		afterEach(() => {
			jest.clearAllMocks();
		});
		it('should update `window.location.pathname` ', () => {
			expect(event.preventDefault).toHaveBeenCalledTimes(1);
		});
	});
	describe('When user clicks on `Link` component and holds `Crtl` key', () => {
		const event = {
			metaKey: false,
			ctrlKey: true,
			preventDefault: jest.fn(),
		};
		beforeEach(() => {
			component.find('a').simulate('click', event);
		});
		afterEach(() => {
			jest.clearAllMocks();
		});
		it('should update `window.location.pathname` ', () => {
			expect(event.preventDefault).toHaveBeenCalledTimes(0);
		});
	});
});
