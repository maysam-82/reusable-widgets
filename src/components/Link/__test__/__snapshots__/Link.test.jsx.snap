// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link component should render Link component 1`] = `
<a
  className="test-class"
  href="/test-route"
  onClick={[Function]}
>
  Test Component
</a>
`;
